---
title: "A2"
subtitle: "Article : STAT3 and GR cooperate to drive gene expression and growth of basal-like triple-negative breast cancer "
author: "Zhaojing Chen"
output: 
  html_document:
    toc : TRUE
    fig_caption: true
bibliography : A2.bib
data: "r Sys.Date()"
link-citations: true
---


```{r, child='MDA cell line normalization.Rmd'}

```

# Description of my data

link to the [data source](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152201) 

paper : Article : STAT3 and GR cooperate to drive gene expression and growth of basal-like triple-negative breast cancer[@conway2020GR]

In the last assignment, I was working with gene expression data  pulling from Geo with id GSE152201. his dataset describes the experiment about  two breast cancer cell lines; HCC70 and MDA. HCC70 is a basal cell line while MDA is a meschemyal cell line. Each cell line was treated with a ethanol control, and a treatment (DEX) with three replicates each. DEX is a inducer of GR, the experiment is trying to find out what genes are unregulated by this inducer. Complemented by other experiments, they ultimately wanted to show that GR coordinates with STAT3 to active gene expression signature for basal-like triple-negative breast cancer.

For this dataset,I filtered the gene that has count per million less than 3 for each cell line, normalizing by treatment using EdgeR. I also mapped the ensembl gene id to HGNC symbol using Biomart and merged the HGNC symbol that was originally existed in my dataset to the unmapped HGNC symbol.The final coverage above 95%

Since this data contains two cell lines, I am just going to perform one differential expression for the MDA cell line.

```{r}
knitr::opts_chunk$set(
  warning = FALSE, message = FALSE,echo = TRUE, results = "hide", tidy = TRUE)
```

# A2 : Differentail Gene Expression 
## Set up

```{r,echo=FALSE}
if (!requireNamespace("BiocManager",quietly = TRUE)){
  install.packages('BioManager')}
library(BiocManager)
if (!requireNamespace("ComplexHeatmap",quietly = TRUE)){
  BiocManager::install('ComplexHeatmap')}
if (!requireNamespace("circlize",quietly = TRUE)){
  install.packages('circlize')}
if (!requireNamespace("limma",quietly = TRUE)){
  BiocManager::install('limma')}
if (!requireNamespace("gprofiler2",quietly = TRUE)){
  BiocManager::install('gprofiler2')}
if (!requireNamespace("tidyverse",quietly = TRUE)){
  install.packages('tidyverse')}
if (!requireNamespace("knitr",quietly = TRUE)){
  install.packages('knitr')}
if (!requireNamespace("DT",quietly = TRUE)){
  install.packages('DT')}
if (!requireNamespace("broom",quietly = TRUE)){
  install.packages('broom')}
if (!requireNamespace("tidyverse",quietly = TRUE)){
  install.packages('tidyverse')}
library(broom)
library(gprofiler2)
library(limma)
library(ComplexHeatmap)
library(circlize)
library(tidyverse)
library(Biobase)

```


All packages references in A2
[@biocmanager;@circlize;@heatmap;@limma;@knitr;@DT;@gprofiler2;@biobase;@tidyverse;@lecture6;@lecture7]



```{r,include=FALSE,eval=FALSE}
# Load the MDA231 cell line data
normalized_count_data <- read.table(file=file.path(getwd(),"data",
                                   "GSE152201_finalized_normalized_counts_mda_2022.txt"),
                                    header = TRUE,sep = "\t",
                                    stringsAsFactors = FALSE, )     

```

```{r}
# rename for convinence
normalized_count_data<-finalized_normalized_counts
# select the normalized data
heatmap_matrix <- normalized_count_data[,
                        3:8]
# rename the column and row names
rownames(heatmap_matrix) <- normalized_count_data$ensembl_gene_id
colnames(heatmap_matrix)
colnames(heatmap_matrix) <- colnames(normalized_count_data[,
                        3:8])

```


## First scaled Heatmap

```{r,eval=FALSE,fig.cap='Figure 1: Initial heatmap for differential expression with row normalization'}

# row normalized the data 
heatmap_matrix <- t(scale(t(heatmap_matrix)))
heatmap_matrix<-data.frame(heatmap_matrix)
# store 104 NAs 
MDA_heatmap_missing <- heatmap_matrix %>% filter_all(any_vars(is.na(.)))
# filter 104 NAs
heatmap_matrix <- heatmap_matrix %>% drop_na(.)
# some NAs, could be due to 0 variance
# select the first 2000?, plot showed nothing
#heatmap_matrix_2000<-heatmap_matrix[1:2000,]

# set up the color gradient
if(min(heatmap_matrix) == 0){
  heatmap_col = colorRamp2(c( 0,max(heatmap_matrix)),
                            c( "white", "red"))
  } else {
  heatmap_col = colorRamp2(c(min(heatmap_matrix), 0,
                             max(heatmap_matrix)),
                             c("blue", "white", "red"))
  }
# plot a heatmap 
current_heatmap <- Heatmap(as.matrix(heatmap_matrix),
                           show_row_dend = TRUE,
                           show_column_dend = TRUE,
                           col=heatmap_col,
                           show_column_names = TRUE,
                           show_row_names = FALSE,
                           show_heatmap_legend = TRUE,row_title = "normalized clustered genes",
                           column_title = "mda_samples")
current_heatmap

```

I did not include this image because the rendering time is long, I have thought about to include the first 2000 but the plot basically showed nothing.

So this initial heatmap  shows there is some separation of signals but is not really clear.

[@heatmap; @circlize; @lecture6]

## T test

Since DEX is an inducer of the GR(NR3C1) gene. I would assume it would be significantly up-regulated. Here I ran a t-test to test this and it is significant
```{r, results='hold'}

# get the column indexes for control and treatment group
etoh_samples <- grep(colnames(normalized_count_data), pattern ="EtoH") 
dex_samples <- grep(colnames(normalized_count_data), pattern ="DEX")

# Find gene of interest, GR(NR3C1)
gr_gene_of_interest <- which(
  normalized_count_data$hgnc_symbol == "NR3C1")

# separate the data into control and treatment sample based on the given column indexes
gr_etoh_samples <-t(normalized_count_data[gr_gene_of_interest,     etoh_samples]) 
colnames(gr_etoh_samples) <-c("etoh_samples") 
#gr_etoh_samples

gr_dex_samples <-t(normalized_count_data[gr_gene_of_interest,     dex_samples]) 
colnames(gr_dex_samples) <-c("dex_samples") 
#gr_dex_samples

# Run a t test
t_test<-t.test(x=t(gr_etoh_samples),y=t(gr_dex_samples))
DT::datatable(broom::tidy(t_test) ,caption  = "Table 1 : T test result for GR(NR3C1)", options = list(scrollX = T))
```

[@broom;@lecture6;@DT]


## Grouping

```{r,results='hold'}
# extract useful information from the column names
samples <- data.frame(
        lapply(colnames(normalized_count_data)[3:8],
        FUN=function(x){
          unlist(strsplit(x,split = "_"))[c(1,2,4)]}))#?
colnames(samples) <- colnames(normalized_count_data)[3:8] 
rownames(samples) <- c("cell_line ", "treatments","replicates") 
samples <- data.frame(t(samples))
knitr::kable(samples[1:6,],output ="html",caption = "Table 2: Showing for ecah sample, which cell line, which replicats, and what condition")


```

[@knitr]

## PCA
The position of control(ETOH) and treatment(DEX) samples on the PCA plot is mostly apart.So I would use a simple model to fit 
```{r}
# 2D representation of my data, serve as reference for model building later
plotMDS(heatmap_matrix,
               col =c(rep("blue",3), rep("red",3)) )
legend("topright",c("ETOH" ,"DEX"),pch = 19,fill = c("blue", "red"))
title(main = " Figure 2 : Principle Component Analysis Plot", sub = "MAD cell line")

```

[@limma;@lecture6]

From the plot we can see there is clear seperation of the two groups, so I will fit a simple model 



## Simple Model


```{r}
# filter duplicated row name (RNA,rRNA)
duplicate_num <-length(normalized_count_data$hgnc_symbol)  -length(unique(normalized_count_data$hgnc_symbol))
# filter duplicated symbil
normalized_count_data<- distinct(normalized_count_data, hgnc_symbol, .keep_all = TRUE) # 15512 rows left

# create a design matrix
model_design <- model.matrix(~ samples$treatments )
knitr::kable(model_design[1:5,], type="html",caption = "model design") # not showm\n in knitted version 
# convert dataframe to matrix
expressionMatrix <- as.matrix(normalized_count_data[,3:8])
# assign row names and column names
rownames(expressionMatrix)<-normalized_count_data$ensembl_gene_id
colnames(expressionMatrix) <- colnames(normalized_count_data)[3:8]
# create a exprssion set 
minimalSet <- ExpressionSet(assayData=expressionMatrix)
# fit the model
fit <- lmFit(minimalSet, model_design)


```

There a `r duplicate_num` duplicated hgnc id in my file, those are likely due to duplicated hgnc id I kept for my first assignment.

```{r,results='hold'}
# use empirical bays to find genes that are differentially expressed
# trend = TRUE is specific for RNA-seq data
fit2 <- eBayes(fit,trend=TRUE)
# get the top-hits after correction for multiple hypothesis testing 
topfit <- topTable(fit2,coef=ncol(model_design),
                   adjust.method = "BH",number = nrow(expressionMatrix))
# merge the hgnc symbol by ensembl id
output_hits <- merge(normalized_count_data[,1:2],topfit,
                     by.y=0,by.x=1,all.y=TRUE)
# order the p value from high to low 
output_hits <- output_hits[order(output_hits$P.Value),]
knitr::kable(output_hits[1:10,2:8],type="html",row.names = FALSE, caption = "Table 4: Simple model fit differential gene expression result using empirica bays")
# number of genes pass un-adjusted threshold
num_p_val<-length(which(output_hits$P.Value < 0.05)) 
# number of genes pass adjusted threshold
num_adj_p <-length(which(output_hits$adj.P.Val < 0.05)) 
# get the fit for my gene of interest
gr<-output_hits[which(output_hits$hgnc_symbol== "NR3C1"),]


DT::datatable(gr,option = list(scrollX=T),caption = "Table 5: model fit result for GR gene")

```
[@DT;@biobase;@limma;@lecture6]



### Q1 and Q2

There are `r num_p_val` genes pass threshold (p value < 0.05) 
There are `r num_adj_p` genes pass correction (adjust p value <0.05)
Since I conducts thousands of statistical test,I used Benjamini-Hochberg(BH) for multiple hypothesis testing correction because BH is not as strict as bonferroni so I would not get all my genes filtered out and it correct false positives. I use 0.05 as my threshold because 0.05 is a general and common p value for statistically analysis.  The purpose of this experiment is to identify all genes that are activated by the addition of DEX. If I do not corrected for mulitple hypothesis testing, there are 4000 more genes up-regulated because of the DEX. This seems a lot to me so I would use the adjusted p value with a threshold of 0.05

## Volcano plot 

### Q3
```{r}
# create a new column to show genes after correction that are up-regulated, down-regulated, and no change

output_hits$diffexpressed <- "no"
output_hits$diffexpressed[output_hits$logFC > 0 & output_hits$adj.P.Val <0.05] <- "up"
output_hits$diffexpressed[output_hits$logFC < 0 & output_hits$adj.P.Val < 0.05] <- "down"
output_hits$diffexpressed[output_hits$hgnc_symbol == "NR3C1"]<-"NR3C1"

# use for plotting to allow my gene of interest to be on the top layer
df_layer_2 <- output_hits[output_hits$diffexpressed=="NR3C1",]

# plot a volcano plot
ggplot()+geom_point(data = output_hits, aes(x=logFC, y=-log10(adj.P.Val),col=diffexpressed))+geom_point(data = df_layer_2, aes(x=logFC, y=-log10(adj.P.Val),col="NR3C1")) + theme_minimal()+xlim(-50,50)+labs(title=" Figure 3: Volcano plot for differential expression")



```

[@ggplot2]

## Heatmap for genes with p value < 0.05

```{r, fig.cap='Figure 4:  heatmap for significant differential expression with row normalization'}
# filter the significant hits
top_hits <- output_hits$ensembl_gene_id[
  output_hits$P.Value<0.05]
# create a heatmap matrix for genes that are significant
heatmap_matrix_tophits <- 
  t(scale(t(heatmap_matrix[which(rownames(heatmap_matrix) %in% top_hits),])))
# set up the color gradient 
if(min(heatmap_matrix_tophits) == 0){
    heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
                             c( "white", "red"))
  } else {
    heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0,
      max(heatmap_matrix_tophits)), c("blue", "white", "red"))
  }
# plot a heatmap  for only those genes that are significant 
current_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = TRUE,
                               show_row_dend = TRUE,
                               show_column_dend = TRUE,
                               col=heatmap_col,
                               show_column_names = TRUE,
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE,
                           row_title = "normalized clustered genes",
                           column_title = "mda_samples"
                               )
current_heatmap

```

[@heatmap; @circlize; @lecture6]



### Q4

After we select for genes that are significant, the heatmap shows a more clear picture of the experiment. 
This is because the experiment might be well-performed to eliminate any variation that is not resulted from the manipulation. It is also shown by the PCA plot that the control and treatment is well-separated.
Gene sets are up-regulated in the control sample are down-regulated in the treatment group. 
Gene sets are down-regulated in the control sample are up-regulated in the treatment group. 



## Ordered heatmap for genes with adjusted p value < 0.05

let us not cluster the genes but group controls and experiments 

```{r,fig.cap='Figure 5:unclustered heatmap for significant differential expression after correction'}
top_hits <- output_hits$ensembl_gene_id[
  output_hits$adj.P.Val<0.05]
heatmap_matrix_tophits <- t(
  scale(t(heatmap_matrix[
    which(rownames(heatmap_matrix) %in% top_hits),])))
if(min(heatmap_matrix_tophits) == 0){
    heatmap_col = colorRamp2(c( 0, max(heatmap_matrix_tophits)),
                             c( "white", "red"))
  } else {
    heatmap_col = colorRamp2(c(min(heatmap_matrix_tophits), 0,
      max(heatmap_matrix_tophits)), c("blue", "white", "red"))
 }


# plot heatmap that orders columns based on what I have given (do not cluster columns)
current_heatmap <- Heatmap(as.matrix(heatmap_matrix_tophits),
                           cluster_rows = TRUE,
                           cluster_columns = TRUE,
                               show_row_dend = TRUE,
                               show_column_dend = FALSE,
                               col=heatmap_col,
                               show_column_names = TRUE,
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE, row_title = "normalized clustered genes",
                           column_title = "mda_samples"
                               )
current_heatmap




```

[@heatmap; @circlize; @lecture6]

There is a  clear clusters of genes if we only inlcude genes that are significant after correction.


# A2: Threshold and over-representation analysis



```{r,include=FALSE,eval=FALSE}

#MDA_output_missing <- output_hits%>% filter_all(any_vars(is.na(.)))

#heatmap_matrix <- heatmap_matrix %>% drop_na(.)

upregulated_genes <- output_hits$hgnc_symbol[which(output_hits$adj.P.Val < 0.05              &output_hits$logFC > 0)] 
downregulated_genes <- output_hits$hgnc_symbol[which(output_hits$adj.P.Val < 0.05              &output_hits$logFC < 0)] 

# create a rank file that contains all genes associated with a rank to define whether the gene is up or down regulated, and how strong is the p value 
output_hits[,"rank"] <- -log10(output_hits$adj.P.Val)*sign(output_hits$logFC)
# order the file my rank 
output_hits<-output_hits[order(output_hits$rank),]
# output up-regulated gene names
write.table(x=upregulated_genes,file=file.path("data","mda_upregulated_genes.txt"),sep = "\t",row.names = FALSE,col.names = FALSE,quote = FALSE)
# output down regulated gene names
write.table(x=downregulated_genes,file=file.path("data","mda_downregulated_genes.txt"),sep = "\t",row.names = FALSE,col.names = FALSE,quote = FALSE)
# output rank file
write.table(x=output_hits[,c("hgnc_symbol","rank")],file=file.path("data","rank.txt"),sep = "\t",row.names = FALSE,col.names = FALSE,quote = FALSE)
```



## Enrichment analysis using gprofiler2

### Q1 
I decided to use the gprofiler2 package for enrichment analysis because it would fit well with the rest of code and gprofiler2 actually come with some nice graphing function
I used Benjamini-Hochberg(BH) for multiple hypothesis testing correction there are thousands of pathways. BH is not as strict as bonferroni so I would not get all pathway out and it corrects false positives.

### Q2
I choose GO biological processes , Reactome and Wikipathway.I choose those becuase I am interested in biological pathway annotation and those three are popular and free. I want my initial analysis to be high quality so I exclude electronic version of GO.

I am using the following version as shown in the [gprofiler web server](https://biit.cs.ut.ee/gprofiler/gost)-Data sources - Show data versions

*GO:BP – annotations: BioMart, classes:releases/2021-12-15

*REAC – annotations: BioMart,classes: 2022-1-3

*WP – 2021-12-10

Some code in this section references a publication about gprofiler2[@gprofiler2].

```{r}
# up-regulated subset
up_subset <- output_hits[which(output_hits$adj.P.Val< 0.05              &output_hits$logFC > 0),] 
# down-regulated subset
down_subset <- output_hits[which(output_hits$adj.P.Val< 0.05              &output_hits$logFC < 0),] 
# whole list
whole <-output_hits[which(output_hits$adj.P.Val<0.05),]
```


```{r,results='hold',fig.dim=c(1,1)}
# enrichment analysis for up regulated subset
gp_up = gost(up_subset$hgnc_symbol, organism = "hsapiens",correction_method = "false_discovery_rate",exclude_iea = TRUE,  sources = c("GO:BP", "REAC", "WP"))
# link to the gprofiler server
gp_up_link= gost(up_subset$hgnc_symbol, organism = "hsapiens",correction_method = "false_discovery_rate",exclude_iea = TRUE,  sources = c("GO:BP", "REAC", "WP"),as_short_link = TRUE)
# filter based on term size and order by p value
gp_up$result->df_up
df_up<-df_up[which(df_up$term_size < 200),]
df_up_ordered <-df_up[order(df_up$p_value),]


up_genesets<-nrow(df_up)

# construct a table about enrichment results
DT::datatable(df_up_ordered[,c(2,3,4,9,10,11)],caption = " Table 6 : Enrichment analysis for up-regulated genes", options = list(scrollX=T))
```

```{r,fig.cap='Figure 6: Enrichment plot for up-regulated genes', fig.dim=c(10,10)}
# plot the enrichment result
p1<-gostplot(gp_up, interactive = FALSE)
# highlight top ten hits
p1_up<-publish_gostplot(p1, highlight_terms = df_up_ordered$term_id[1:10])

# publish_gosttable(gp_up)

```

```{r,results='hold'}

# for down regulated subset
gp_down = gost(down_subset$hgnc_symbol, organism = "hsapiens",correction_method = "false_discovery_rate",exclude_iea = TRUE,  sources = c("GO:BP", "REAC", "WP"))
#link
gp_down_link = gost(down_subset$hgnc_symbol, organism = "hsapiens",correction_method = "false_discovery_rate",exclude_iea = TRUE,  sources = c("GO:BP", "REAC", "WP"),as_short_link = TRUE)

# filter based on term size and order by p value
gp_down$result->df_down
df_down<-df_down[which(df_down$term_size < 200),]
df_down_ordered <-df_down[order(df_down$p_value),]

down_genesets<-nrow(df_down)

# construct a table about enrichment results
DT::datatable(df_down_ordered[,c(2,3,4,9,10,11)],caption = " Table 7 : Enrichment analysis for down-regulated genes", options = list(scrollX=T))

```

```{r, fig.cap='Figure 7: Enrichment plot for down-regulated genes',fig.dim=c(10,10)}


# plot the enrichment result
p2<-gostplot(gp_down, interactive = FALSE)
# highlight top ten hits
p2_down<-publish_gostplot(p2, highlight_terms = df_down_ordered$term_id[1:10])

```

```{r,results='hold'}
# for the whole set 
gp_whole = gost(whole$hgnc_symbol, organism = "hsapiens",correction_method = "false_discovery_rate",exclude_iea = TRUE,  sources = c("GO:BP", "REAC", "WP"))

gp_whole_link = gost(whole$hgnc_symbol, organism = "hsapiens",correction_method = "false_discovery_rate",exclude_iea = TRUE,  sources = c("GO:BP", "REAC", "WP"), as_short_link = TRUE)

# filter based on term size and order by p vlaue 
gp_whole$result->df_whole
df_whole<-df_whole[which(df_whole$term_size < 200),]
df_whole_ordered<-df_whole[order(df_whole$p_value),]

whole_genesets<-nrow(df_whole)
# construct a table about enrichment results
DT::datatable(df_whole_ordered[,c(2,3,4,9,10,11)],caption = " Table 8 : Enrichment analysis for down-regulated genes", options = list(scrollX=T))


```

```{r,fig.dim=c(10,10), 'Figure 8: Enrichment plot for all genes'}
# plot the enrichment result
p3<-gostplot(gp_whole, interactive = FALSE)
# highlight top ten hits
p3_whole<-publish_gostplot(p3, highlight_terms = df_whole_ordered$term_id[1:10])

```

[@gprofiler2;@DT]

I have included three links to the gprofiler web server my enrichment analysis results.

up : `r gp_up_link`

down : `r gp_down_link`

whole : `r gp_whole_link`

### Q3
For term size between 0 and 200, threshold of 0.05 after BH correction. There are `r up_genesets` up-regulated genesets. There are `r down_genesets` down-regulated genesets. There are a total  `r whole_genesets` genesets using the whole list. 

### Q4
Comparing enrichment analysis results from using only up or down-regulated gene list versus using the whole list I see there are some differences in the top hits returned. Using the whole gene list gave us some new enrichment results such as MAPK cascade and miRNA transcription. MAPK cascade and miRNA all have regulatory roles in breast cancer progression[@mapk ;@mirna]. This is likely due the reason that some particular genes are highly expressed while other genes in the same geneset  were not. However, in combination they gave a stronger signal. 



# Interpretation
### Q1 and Q2
Yes, the enrichment analysis for up regulated genes supports the result of the paper that GR  actives gene expression signature for basal-like triple-negative breast cancer. This is because among the 10 significant genesets in the enrichment analysis, NF-kB stimulates proliferation and stops programmed cell death in many cell types including basal breast cancer cells[@NFkB]. Also,receptor tyrosine kinases is an important signalling pathway that is involved in breast cancer[@rtk].

Because MDA is a stem cell line. There are some other top hits that are related to stem cell characteristics. 
However, because here I just analyzed differential expression after the induction of GR not STAT3, My analysis cannot confidently support the final conclusion of the paper that STAT3 and GR coordinates to active  basal-like triple-negative breast cancer.

### Journal link
[A2](https://github.com/bcb420-2022/Zhaojing_Chen.git)



# Citations 




